### ISSUE TITLE
[`Microsoft.ML.OnnxRuntime.*` NuGet Packages] Replace Xamarin.iOS + Xamarin.Android Dependencies with .NET iOS and .NET Android

### ISSUE THREAD

**Issue by brminnick** [microsoft/onnxruntime#11618]
## The Problem

The `Microsoft.ML.OnnxRuntime.*` NuGet Packages do not currently support  the new .NET 6.0 mobile libraries:
- `net6.0-ios`
- `net6.0-android`

These two new libraries were introduced in .NET 6.0 and they deprecate Xamarin.iOS and Xamarin.Android, respectively.

.NET 6.0 also introduced .NET MAUI which deprecates Xamarin.Forms.

Going forward, all future C# Mobile Apps will be written using the `net6.0-ios` and `net6.0-android` libraries, and most will leverage .NET MAUI, the cross-platform UI Framework that runs on the `net6.0-ios` and `net6.0-android` libraries.

tl;dr - .NET MAUI Developers cannot consume `Microsoft.ML.OnnxRuntime.*` NuGet Packages until we publish a new release/update that supports `net6.0-ios` and `net6.0-android`

## Current Onnx Support of Xamarin.iOS + Xamarin.Android

The [Microsoft.ML.OnnxRuntime NuGet Package](https://www.nuget.org/packages/Microsoft.ML.OnnxRuntime/) currently supports Xamarin.iOS + Xamarin.Android (aka MonoAndroid). 

The **Dependencies** tab on its [NuGet page](https://www.nuget.org/packages/Microsoft.ML.OnnxRuntime/), specifically, shows its support for Xamarin.iOS 1.0 and MonoAndroid 11.0 (aka Xamarin.Android):

<img width="568" alt="image" src="https://user-images.githubusercontent.com/13558917/170148110-ba1c7d00-0797-4c66-9f4b-9e5a619fb93f.png">

These NuGet Packages are incompatible for use in `net6.0-ios` and `net6.0-android` apps.

# The Solution

The solution is actually pretty "easy".

In the `Microsoft.ML.OnnxRuntime.csproj`, we just need to replace `xamarinios10;monoandroid11.0` with `net6.0-ios` and `net6.0-android`, and add `net6.0`:
https://github.com/microsoft/onnxruntime/blob/e3a2d5cca8bcefe064f83d57e46ea51ddb2b16e8/csharp/src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj#L12

becomes 

```xml
<TargetFrameworks>netstandard1.1;netstandard2.0;net6.0-ios;net6.0-android;net5.0;net6.0;netcoreapp3.1</TargetFrameworks> 
```

After forking the repo and making the above change, I was able to build + compile `Microsoft.ML.OnnxRuntime.csproj` successfully with no build errors.

However, nothing is that easy....

## Tests + Samples

`OnnxRuntime.CSharp.sln` includes the following tests and samples that also need to be ported from Xamarin.Forms to .NET MAUI:
- Test
  - Microsoft.ML.OnnxRuntime.Tests.iOS
  - Microsoft.ML.OnnxRuntime.Tests.Droid
- Sample 
  - Microsoft.ML.OnnxRuntime.InferenceSample.Forms
  - Microsoft.ML.OnnxRuntime.InferenceSample.Forms.Droid
  - Microsoft.ML.OnnxRuntime.InferenceSample.Forms.iOS

### Tests
The Tests projects run UI Tests to validate the ONNX libraries running inside a Xamarin.Forms app.

Unfortunately, they will require a near re-write because they rely upon the following NuGet Packages which have no existing road map to add support for the .NET 6.0 Mobile Libraries:
- Xamarin.TestCloud.Agent
- unit.runner.devices

These UI Test projects will need to leverage a .NET MAUI UI Test framework, and I am not aware of any that currently exist

### Samples
The Samples projects will be a bit easier to update as .NET MAUI is (nearly) identical to Xamarin.Forms. 

To port the sample from Xamarin to MAUI, I recommend the following steps:
1.  Create a new app using the .NET MAUI Template
  - In the CLI, enter `dotnet new maui -n Microsoft.ML.OnnxRuntime.InferenceSample.Maui`
2.  Copy/paste the Xamarin.Forms code into the newly created .NET MAUI library
3. Change the namespaces from `Xamarin.Forms.*` to `Microsoft.Maui.*`

### Resources for Porting Xamarin.Forms libraries to .NET MAUI

We have a few guest blog posts written by Microsoft MVPs where they've shared their experience and their tips after porting their libraries to .NET MAUI:
- [Migrating MR.Gestures from Xamarin.Forms to .NET MAUI](https://devblogs.microsoft.com/xamarin/migrating-mrgestures-to-dotnet-maui/?WT.mc_id=dotnet-0000-bramin)
- [Tips for Moving Your Xamarin Library to .NET MAUI](https://devblogs.microsoft.com/xamarin/tips-for-porting-your-xamarin-library-to-dotnet-maui/?WT.mc_id=dotnet-0000-bramin)

**Comment by brminnick** [comment]
As a fellow Microsoft FTE (and former Xamarin employee), I'm happy to jump on a call to help walk the team through this!

**Comment by skottmckay** [comment]
Thanks for the info and offer of help! Much appreciated.

Couple of questions:

If we replace the xamarin target frameworks does that mean someone who has not updated to MAUI can no longer use the nuget package with their Xamarin app? If so, is it possible to add the .net6 target frameworks to the existing lists and support both Xamarin and MAUI? 

Do we also need updates to the csproj to set things like this, which we use in a few places for doing some platform specific stuff at compile time? 

https://github.com/microsoft/onnxruntime/blob/147a1737f932a093cc05bb029c86c3f5b0add592/csharp/src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.csproj#L93-L99

e.g. that's used here, and I'm not aware of a way where we could do that programmatically at runtime)

https://github.com/microsoft/onnxruntime/blob/147a1737f932a093cc05bb029c86c3f5b0add592/csharp/src/Microsoft.ML.OnnxRuntime/AssemblyInfo.shared.cs#L1-L5

@mikeparker104 was the test project using Xamarin.TestCloud.Agent primarily for enabling testing via AppCenter?

**Comment by mikeparker104** [comment]
@skottmckay Xamarin.UITest was used to support local and automated testing (via App Center). They're still valid for the Xamarin.Forms target but would not currently support MAUI for the reasons outlined by @brminnick. In terms of the TargetGrameworks, it should be possible to support Xamarin.Forms and MAUI in parallel

**Comment by brminnick** [comment]
> If we replace the xamarin target frameworks does that mean someone who has not updated to MAUI can no longer use the nuget package with their Xamarin app? If so, is it possible to add the .net6 target frameworks to the existing lists and support both Xamarin and MAUI?

Yes, and yes. Apologies for implying we have to **replace** the Xamarin.Forms targets; we can keep it and add `net6.0-ios`, net6.0-android` and `net6.0`.

**Comment by vpenades** [comment]
I've just started a new Android project using net6.0-android and I've been blocked by this issue.

Is there a timeline to support net6-android? (preview nugets?)

It is possible to use current nuget packages in a net6.0-android csproj, using some kind of retargeting/compatibility tricks?

**Comment by skottmckay** [comment]
Should be supported in next release. You could try a pre-release build from https://aiinfra.visualstudio.com/PublicPackages/_artifacts/feed/ORT-Nightly

**Comment by vpenades** [comment]
@skottmckay I've tried using the nightly builds, but just adding the nugets into a net6.0-android app produces several errors related to `IOnnxValue` interface.

**Comment by skottmckay** [comment]
> @skottmckay Scott McKay FTE I've tried using the nightly builds, but just adding the nugets into a net6.0-android app produces several errors related to `IOnnxValue` interface.

There is a recent change that might not have made it in to prevent VS auto-generating invalid bindings from the Android AAR (which are unnecessary when we have a package with C# bindings).

You can add this temporarily to work around the problem.

```C#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AI.Onnxruntime
{
    public partial class OnnxMap : Java.Lang.Object, IOnnxValue
    {
        IValueInfo IOnnxValue.Info
        {
            get => Info;
        }

        Java.Lang.Object IOnnxValue.Value
        {
            get => (Java.Lang.Object)Value;
        }
    }

    public partial class OnnxSequence : Java.Lang.Object, IOnnxValue
    {
        IValueInfo IOnnxValue.Info
        {
            get => Info;
        }

        Java.Lang.Object IOnnxValue.Value
        {
            get => (Java.Lang.Object)Value;
        }
    }

    public partial class OnnxTensor : Java.Lang.Object, IOnnxValue
    {
        IValueInfo IOnnxValue.Info
        {
            get => Info;
        }
    }
}
```

**Comment by skottmckay** [comment]
These packages should have the change:
https://aiinfra.visualstudio.com/PublicPackages/_artifacts/feed/ORT-Nightly/NuGet/Microsoft.ML.OnnxRuntime/overview/1.12.0-dev-20220707-0901-62226d030
https://aiinfra.visualstudio.com/PublicPackages/_artifacts/feed/ORT-Nightly/NuGet/Microsoft.ML.OnnxRuntime.Managed/overview/1.12.0-dev-20220707-0901-62226d030

**Comment by skottmckay** [comment]
Added to 1.12 release

**Comment by juwens** [comment]
this happens still with 1.12 and 1.13
